File: ios/OakApp/Features/Usage/UsageDexLink.swift
import Foundation

/// Usage-species list sections. Only moves, items, abilities, and teammates
/// have a Dex page; natures and EV spreads do not.
enum UsageListKind: String, Sendable {
  case moves
  case items
  case abilities
  case natures
  case spreads
  case teammates

  var dexKind: EntityKind? {
    switch self {
    case .moves: return .move
    case .items: return .item
    case .abilities: return .ability
    case .teammates: return .pokemon
    case .natures, .spreads: return nil
    }
  }
}

/// Maps a usage-share row onto a Dex entity route. Pure — no I/O.
enum UsageDexLink {
  /// `nil` when the section has no Dex page or the name is blank after trim.
  static func route(kind: UsageListKind, name: String) -> DexEntityRoute? {
    let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty, let entityKind = kind.dexKind else { return nil }
    return DexEntityRoute(kind: entityKind, query: trimmed)
  }

  /// Header “View in Dex” for the species being inspected.
  static func speciesRoute(nameOrSlug: String) -> DexEntityRoute? {
    DexEntityRoute.pokemonPage(named: nameOrSlug)
  }
}

File: android/app/src/main/kotlin/ai/gowtam/oak/features/usage/UsageDexLink.kt
package ai.gowtam.oak.features.usage

import ai.gowtam.oak.wire.EntityKind

/**
 * Usage-species list sections. Only moves, items, abilities, and teammates
 * have a Dex page; natures and EV spreads do not.
 */
enum class UsageListKind {
    MOVES,
    ITEMS,
    ABILITIES,
    NATURES,
    SPREADS,
    TEAMMATES,
    ;

    val dexKind: EntityKind?
        get() = when (this) {
            MOVES -> EntityKind.MOVE
            ITEMS -> EntityKind.ITEM
            ABILITIES -> EntityKind.ABILITY
            TEAMMATES -> EntityKind.POKEMON
            NATURES, SPREADS -> null
        }
}

data class UsageDexTarget(
    val kind: EntityKind,
    val query: String,
)

/** Maps a usage-share row onto a Dex entity. Pure — no I/O. */
object UsageDexLink {
    /** `null` when the section has no Dex page or the name is blank after trim. */
    fun route(kind: UsageListKind, name: String): UsageDexTarget? {
        val trimmed = name.trim()
        val entityKind = kind.dexKind
        if (trimmed.isEmpty() || entityKind == null) return null
        return UsageDexTarget(entityKind, trimmed)
    }

    /** Header “View in Dex” for the species being inspected. */
    fun speciesRoute(nameOrSlug: String): UsageDexTarget? {
        val trimmed = nameOrSlug.trim()
        if (trimmed.isEmpty()) return null
        return UsageDexTarget(EntityKind.POKEMON, trimmed)
    }
}
